Sources/Basic/TerminalController.swift

```diff
@@ -0,0 +1,252 @@
+        // Return the original line if we don't need to truncate.
+        guard !dumb && line.utf8.count > terminalWidth else {
+            return line
+        }
+        let mid = terminalWidth/2
+        let frontIndex = line.index(line.startIndex, offsetBy: max(0, mid - 2))
+        let backIndex = line.index(line.endIndex, offsetBy: -(mid + 1))
+        return line[line.startIndex..<frontIndex] + "..." + line[backIndex..<line.endIndex]
+    }
+
+    /// Wraps the string into the color mentioned. If terminal is dumb no color code will be inserted.
+    private func wrap(_ line: String, inColor color: TermColor) -> String {
+        guard !dumb || !line.isEmpty || color == .noColor else {
+            return line
+        }
+        return color.string + line + resetString
+    }
+}
+
+// FIXME: Move progessbar to utility?
+/// A protocol to operate on terminal based progress bars.
+public protocol ProgressBarProtocol {
+    func update(percent: Int, text: String)
+}
+
+/// Simple ProgressBar which shows the update text in new lines.
+public final class SimpleProgressBar<Target: OutputStream>: ProgressBarProtocol {
+    private let term: TerminalController<Target>
+    private let header: String
+    private var isClear: Bool
+
+    public init(term: TerminalController<Target>, header: String) {
+        self.term = term
+        self.header = header
+        self.isClear = true
+    }
+
+    public func update(percent: Int, text: String) {
+        if isClear {
+            term.writeLine(line: header)
+            term.endLine()
+            isClear = false
+        }
+
+        term.writeLine(line: "\(percent)%: " + text)
+        term.endLine()
+    }
+}
+
+/// Three line progress bar with header, redraws on each update.
+final class ProgressBar<Target: OutputStream>: ProgressBarProtocol {
+    private let term: TerminalController<Target>
+    private let header: String
+    private var isClear: Bool // true if haven't drawn anything yet.
+
+    public init(term: TerminalController<Target>, header: String) {
+        self.term = term
+        self.header = header
+        self.isClear = true
+    }
+
+    func update(percent: Int, text: String) {
+        if isClear {
+            term.writeLine(line: header, color: .red)
+            term.endLine()
+            isClear = false
+        }
+
+        term.clearline()
+        let percentString = percent < 10 ? "0\(percent)" : "\(percent)"
+        let prefix = "\(percentString)% ["
+        term.write(string: prefix, color: .green)
+
+        // FIXME: Fix the calculation here.
+        let barWidth = terminalWidth - prefix.utf8.count - 20
+        let n = Int(Double(barWidth) * Double(percent)/100.0)
+
+        term.write(string: "=".repeating(n: n) + "-".repeating(n: barWidth - n))
+        term.write(string: "]", color: .green)
+        term.endLine()
+
+        term.clearline()
+        term.writeLine(line: text)
+
+        if percent >= 100 {
+            term.endLine()
```

Comment:
This doesn't seem like the right place to embed this, I would expect a separate method.